### PR TITLE
[Fix #710] Use network_id to identify faucet_url

### LINKED ISSUE #710: Use Network ID to determine the correct faucet to use
Code reference: https://github.com/XRPLF/xrpl-py/blob/main/xrpl/asyncio/wallet/wallet_generation.py

Currently, the faucet generation code guesses the correct faucet to use based on substrings in the URL. This isn't a very robust system. It would be better (and more robust) to use the Network ID.

### PR DESCRIPTION
**Pull request by ckeshava**
## High Level Overview of Change
- use network_id to disambiguate between different XRPL chains
- move the utility function _get_network_id_and_build_version into client.py file for convinience

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (non-breaking change that only restructures code)
- [x] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### Did you update CHANGELOG.md?

- [ ] Yes
- [x] No, this change does not impact library users
This change should not modify the experience for external developers

## Test Plan

Tests have been modified to accomodate the new API for grabbing the faucet_url

### CODE REVIEW

File: xrpl/asyncio/clients/client.py [hunk 1]
```diff
@@ -51,3 +54,24 @@ async def _request_impl(
         :meta private:
         """
         pass
+
+
+async def _get_network_id_and_build_version(client: Client) -> None:
```
Comment (mvadari) [inline xrpl/asyncio/clients/client.py:59]: Why not automatically run this on every client upon instantiation, instead of having it be a helper function? Then it only needs to be run once, instead of every time a function that uses this is called.

Comment (ckeshava) [reply xrpl/asyncio/clients/client.py:59]: We are not duplicating invocations of this function. Here are [two](https://github.com/XRPLF/xrpl-py/blob/4779d796e6f808c952214e794be5d9d342412e14/xrpl/asyncio/transaction/main.py#L233) [examples](https://github.com/XRPLF/xrpl-py/blob/ea6365af1b3da29c45f4441fefcfb298f92d897d/xrpl/asyncio/wallet/wallet_generation.py#L62) where the `network_id` information persists in the client object.

But I agree, this could be done in a cleaner way.
I can only invoke `_get_network_id_and_build_version` function after a client has opened the socket-connection. So, I'll need to invoke this in this `open()` method of all the child classes of `Client`

let me try it out

Comment (mvadari) [reply xrpl/asyncio/clients/client.py:59]: I'm open to the option of doing it lazily - i.e. only if and as needed. All that would require is checking if the network ID has been instantiated before running the request.

Comment (ckeshava) [reply xrpl/asyncio/clients/client.py:59]: okay, I've added a check: [23d3eb3](https://github.com/XRPLF/xrpl-py/pull/715/commits/23d3eb31df830a1c55f2cae51e720fde14c673a3)

File: xrpl/asyncio/transaction/main.py [hunk 2]
```diff
@@ -1,17 +1,19 @@
 """High-level transaction methods with XRPL transactions."""
+
 import math
 from typing import Any, Dict, Optional, cast
 
 from typing_extensions import Final
 
 from xrpl.asyncio.account import get_next_valid_seq_number
 from xrpl.asyncio.clients import Client, XRPLRequestFailureException
+from xrpl.asyncio.clients.client import _get_network_id_and_build_version
```
Comment (mvadari) [inline xrpl/asyncio/transaction/main.py:10]: Any function that gets imported elsewhere shouldn't start with a `_`.

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 3]
```diff
@@ -57,7 +59,9 @@ async def generate_faucet_wallet(
 
     .. # noqa: DAR402 exception raised in private method
     """
-    faucet_url = get_faucet_url(client.url, faucet_host)
+    if not client.network_id:
+        await _get_network_id_and_build_version(client)
```
Comment (mvadari) [inline xrpl/asyncio/wallet/wallet_generation.py:63]:
```suggestion
    await _get_network_id_and_build_version(client)
```
It's already in the helper function.

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 4]
```diff
@@ -150,34 +154,40 @@ def process_faucet_host_url(input_url: str) -> str:
...
 
     Returns:
         The URL of the matching faucet.
 
     Raises:
-        XRPLFaucetException: if the provided URL is not for the testnet or devnet.
+        XRPLFaucetException: if the provided network_id does not correspond to testnet
+            or devnet.
     """
+    if network_id is not None:
```
Comment (mvadari) [inline xrpl/asyncio/wallet/wallet_generation.py:176]: This check shouldn't be needed (or should throw an error) because the network ID should never be `None`.

Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py:175]: well, network_id is `None` in some cases -- the rippled integration tests do not specify a `network_id` in the config file, hence this value is `None`

I've updated the config file in one of the recent xrpl-py PRs to fix this issue. Nevertheless, I don't think it's good idea to throw an error if `network_id == None`

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 5]
```diff
@@ -150,34 +154,40 @@ def process_faucet_host_url(input_url: str) -> str:
...
+            return _TEST_FAUCET_URL
+        elif network_id == 2:
+            return _DEV_FAUCET_URL
+        elif network_id == 0:
+            raise XRPLFaucetException("Cannot create faucet with a client on mainnet.")
+        elif network_id < 0:
+            raise XRPLFaucetException("network_id cannot be negative.")
+
     if faucet_host is not None:
         return process_faucet_host_url(faucet_host)
```
[the remark this answers is not in the document]
Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py]: 👍  fixed in  ab918dc

File: xrpl/asyncio/transaction/main.py (hunk 2 above)
Comment (ckeshava) [reply xrpl/asyncio/transaction/main.py]: fixed in ab918dc

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 6]
```diff
@@ -150,34 +154,40 @@ def process_faucet_host_url(input_url: str) -> str:
...
+
+    if network_id is not None:
+        if network_id == 1:
+            return _TEST_FAUCET_URL
+        elif network_id == 2:
+            return _DEV_FAUCET_URL
+        elif network_id == 0:
+            raise XRPLFaucetException("Cannot create faucet with a client on mainnet.")
+        elif network_id < 0:
+            raise XRPLFaucetException("network_id cannot be negative.")
```
Comment (mvadari) [inline xrpl/asyncio/wallet/wallet_generation.py:187]: Does rippled allow this scenario?

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 7]
```diff
@@ -150,34 +154,40 @@ def process_faucet_host_url(input_url: str) -> str:
...
+
     raise XRPLFaucetException(
-        "Cannot fund an account with a client that is not on the testnet or devnet."
+        "Cannot create faucet URL without network_id or the faucet_host information."
     )
```
Comment (mvadari) [inline xrpl/asyncio/wallet/wallet_generation.py:191]: Thoughts on throwing this error in the main `generate_faucet_wallet` function instead of here? e.g.
```python
if network_id is None and faucet_host is None:
  raise XRPLFaucetException(...)
```

Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py:191]: I don't object to that idea, but how is that different from the current implementation?

At line 189 (the last line of the `get_faucet_url` method), we have established that `faucet_host == None`.  You are correct, an unknown devnet might have a networkID, but how can I determine the host_url from a networkID (say `1345` ?)

Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py:191]: > ```python
> if network_id is None and faucet_host is None:
>   raise XRPLFaucetException(...)
> ```
The said devnet will have a networkID, hence this condition needs to be:
```python
if faucet_host is None and networkID >= 1025:
  raise XRPLFaucetException(...)
```

The current implementation is equivalent to the above if-condition

Comment (mvadari) [reply xrpl/asyncio/wallet/wallet_generation.py:191]: The difference comes from typing and failing fast. IMO it makes more sense for `get_faucet_url` to not accept `None` as a valid network ID.

You can't determine the host URL from an unknown Devnet, but the error message should be different.

File: xrpl/asyncio/wallet/wallet_generation.py (hunk 6 above)
Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py:187]: the config file does not allow negative numbers as input for network_id. But it accepts really large numbers (for eg: `4294967295` which could be negative numbers, if typecast into smaller signed-types)

also, rippled unit tests allow negative numbers - https://github.com/XRPLF/rippled/compare/develop...ckeshava:rippled:negNetworkID

Comment (mvadari) [reply xrpl/asyncio/wallet/wallet_generation.py:187]: I don't think xrpl-py needs any specific checks for this, "we don't understand this network ID" is fine.

File: xrpl/asyncio/wallet/wallet_generation.py (hunk 7 above)
Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py:191]: we cannot expect all networks to have a valid `NetworkID` isn't it? 
If we force `get_faucet_url` to accept valid networkIDs only, then some of the processing logic shifts into `generate_faucet_wallet` method.

```
def get_faucet_url(network_id: int) -> str:
    if network_id == 1:
        return _TEST_FAUCET_URL
    elif network_id == 2:
        return _DEV_FAUCET_URL
    elif network_id == 0:
        raise XRPLFaucetException("Cannot create faucet with a client on mainnet.")
    elif network_id < 0:
        raise XRPLFaucetException("network_id cannot be negative.")

    # unreachable code
    return ""

async def generate_faucet_wallet(
    client: Client,
    wallet: Optional[Wallet] = None,
    debug: bool = False,
    faucet_host: Optional[str] = None,
    usage_context: Optional[str] = None,
    user_agent: Optional[str] = "xrpl-py",
) -> Wallet:
    if not client.network_id:
        await get_network_id_and_build_version(client)

    if client.network_id is None and faucet_host is None:
        raise XRPLFaucetException(
            "Cannot create faucet URL without network_id or the faucet_host information"
        )

    if faucet_host is not None:
        faucet_url = process_faucet_host_url(faucet_host)
    else:
        faucet_url = get_faucet_url(client.network_id)
        
    # remaining code in the function ...

```

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 8]
```diff
@@ -150,34 +154,40 @@ def process_faucet_host_url(input_url: str) -> str:
...
     Returns:
         The URL of the matching faucet.
 
     Raises:
-        XRPLFaucetException: if the provided URL is not for the testnet or devnet.
+        XRPLFaucetException: if the provided network_id does not correspond to testnet
+            or devnet.
     """
     if faucet_host is not None:
         return process_faucet_host_url(faucet_host)
```
[the remark this answers is not in the document]
Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py:177]: yeah, I put down a snippet in a comment below

File: xrpl/asyncio/wallet/wallet_generation.py (hunk 7 above)
Comment (mvadari) [reply xrpl/asyncio/wallet/wallet_generation.py:191]: That flow makes more sense to me, personally.

Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py:191]: alrighty

File: tests/unit/asyn/wallet/test_wallet.py [hunk 9]
```diff
@@ -16,30 +17,16 @@ def test_wallet_get_xaddress(self):
...
-    def test_get_faucet_wallet_custom(self):
-        json_client_url = "https://s.devnet.rippletest.net:51234"
-        ws_client_url = "wss://s.devnet.rippletest.net/"
-        custom_host = "my_host.org"
-        expected_faucet = "https://my_host.org/accounts"
-
-        self.assertEqual(get_faucet_url(json_client_url, custom_host), expected_faucet)
-        self.assertEqual(get_faucet_url(ws_client_url, custom_host), expected_faucet)
-
-    def test_get_faucet_wallet_test(self):
```
Comment (ckeshava) [inline tests/unit/asyn/wallet/test_wallet.py:36]: these test cases are already covered in the below class `TestProcessFaucetHostURL`

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 10]
```diff
@@ -150,35 +162,32 @@ def process_faucet_host_url(input_url: str) -> str:
...
-    raise XRPLFaucetException(
-        "Cannot fund an account with a client that is not on the testnet or devnet."
-    )
+    elif network_id == 0:
+        raise XRPLFaucetException("Cannot create faucet with a client on mainnet.")
+    elif network_id < 0:
+        raise XRPLFaucetException("Unable to parse the specified network_id.")
+
+    # this line is unreachable. Custom devnets must specify a faucet_host input
+    return ""
```
Comment (mvadari) [inline xrpl/asyncio/wallet/wallet_generation.py:190]: You should raise an error here.

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 11]
```diff
@@ -150,35 +162,32 @@ def process_faucet_host_url(input_url: str) -> str:
...
-    if faucet_host is not None:
-        return process_faucet_host_url(faucet_host)
-    if "altnet" in url or "testnet" in url:  # testnet
+    if network_id == 1:
         return _TEST_FAUCET_URL
-    if "sidechain-net2" in url:  # sidechain issuing chain devnet
-        raise XRPLFaucetException(
-            "Cannot fund an account on an issuing chain. Accounts must be created via "
-            "the bridge."
-        )
```
Comment (mvadari) [inline xrpl/asyncio/wallet/wallet_generation.py:176]: This piece still needs to be included. The network ID is 262.

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 12]
```diff
@@ -150,35 +162,32 @@ def process_faucet_host_url(input_url: str) -> str:
...
 
     Raises:
-        XRPLFaucetException: if the provided URL is not for the testnet or devnet.
+        XRPLFaucetException: if the provided network_id does not correspond to testnet
+            or devnet.
     """
-    if faucet_host is not None:
-        return process_faucet_host_url(faucet_host)
-    if "altnet" in url or "testnet" in url:  # testnet
+    if network_id == 1:
```
Comment (mvadari) [inline xrpl/asyncio/wallet/wallet_generation.py:180]: Can this if-else be replaced with a dictionary? It'll be easier to maintain that way.

Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py]: updated in [f06e5c1](https://github.com/XRPLF/xrpl-py/pull/715/commits/f06e5c15e618816ee7478909695f37a38618de1d)

File: xrpl/asyncio/wallet/wallet_generation.py (hunk 11 above)
Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py]: thanks. fixed in [f06e5c1](https://github.com/XRPLF/xrpl-py/pull/715/commits/f06e5c15e618816ee7478909695f37a38618de1d)

File: xrpl/asyncio/wallet/wallet_generation.py (hunk 10 above)
Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py]: fixed in [f06e5c1](https://github.com/XRPLF/xrpl-py/pull/715/commits/f06e5c15e618816ee7478909695f37a38618de1d)

File: xrpl/asyncio/wallet/wallet_generation.py (hunk 6 above)
Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py]: updated in [f06e5c1](https://github.com/XRPLF/xrpl-py/pull/715/commits/f06e5c15e618816ee7478909695f37a38618de1d)

File: xrpl/asyncio/wallet/wallet_generation.py (hunk 7 above)
Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py]: updated sometime before [f06e5c1](https://github.com/XRPLF/xrpl-py/pull/715/commits/f06e5c15e618816ee7478909695f37a38618de1d)

File: xrpl/asyncio/wallet/wallet_generation.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ckeshava) [reply xrpl/asyncio/wallet/wallet_generation.py]: updated somewhere prior to [f06e5c1](https://github.com/XRPLF/xrpl-py/pull/715/commits/f06e5c15e618816ee7478909695f37a38618de1d)

File: CHANGELOG.md [hunk 13]
```diff
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 - Allow empty strings for the purpose of removing fields in DIDSet transaction
+- Refactor the API of `get_faucet_wallet` method. This produces a non-ambiguous URL for Testnet and Devnet faucet hosts.
```
Comment (mvadari) [inline CHANGELOG.md:15]:
```suggestion
- Use `NetworkID` in faucet processing to produce a non-ambiguous URL for faucet hosts.
```

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 14]
```diff
@@ -150,34 +162,40 @@ def process_faucet_host_url(input_url: str) -> str:
...
-        XRPLFaucetException: if the provided URL is not for the testnet or devnet.
+        XRPLFaucetException: if the provided network_id does not correspond to testnet
+            or devnet.
     """
-    if faucet_host is not None:
-        return process_faucet_host_url(faucet_host)
-    if "altnet" in url or "testnet" in url:  # testnet
-        return _TEST_FAUCET_URL
-    if "sidechain-net2" in url:  # sidechain issuing chain devnet
+    map_network_id_to_url: Dict[int, str] = {1: _TEST_FAUCET_URL, 2: _DEV_FAUCET_URL}
```
Comment (mvadari) [inline xrpl/asyncio/wallet/wallet_generation.py:180]: This object should be outside of the function, since it's not dependent on the data inside. And the name should be in uppercase.

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 15]
```diff
@@ -150,34 +162,40 @@ def process_faucet_host_url(input_url: str) -> str:
...
-        return _DEV_FAUCET_URL
+    elif network_id == 0:
+        raise XRPLFaucetException("Cannot create faucet with a client on mainnet.")
+    elif network_id < 0:
+        raise XRPLFaucetException("Unable to parse the specified network_id.")
+
+    # this line is unreachable. Custom devnets must specify a faucet_host input
     raise XRPLFaucetException(
-        "Cannot fund an account with a client that is not on the testnet or devnet."
+        "The specified network_id " + str(network_id) + " is not recognizable"
```
Comment (mvadari) [inline xrpl/asyncio/wallet/wallet_generation.py:198]:
```suggestion
        "The specified network_id " + str(network_id) + " does not have a known faucet."
```

File: xrpl/asyncio/wallet/wallet_generation.py [hunk 16]
```diff
@@ -17,6 +17,7 @@
 _DEV_FAUCET_URL: Final[str] = "https://faucet.devnet.rippletest.net/accounts"
 
 _TIMEOUT_SECONDS: Final[int] = 40
+_MAP_NETWORK_ID_TO_URL: Dict[int, str] = {1: _TEST_FAUCET_URL, 2: _DEV_FAUCET_URL}
```
Comment (mvadari) [inline xrpl/asyncio/wallet/wallet_generation.py:20]: nit:
```suggestion
_NETWORK_ID_URL_MAP: Dict[int, str] = {1: _TEST_FAUCET_URL, 2: _DEV_FAUCET_URL}
```

**Review by mvadari — APPROVED** [review]

**Review by anissa-ripple — APPROVED** [review]
